src/main/java/org/jabref/gui/preferences/PreferencesSearchHandler.java

```diff
@@ -0,0 +1,140 @@
+package org.jabref.gui.preferences;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+
+import javafx.beans.property.ListProperty;
+import javafx.beans.property.SimpleListProperty;
+import javafx.beans.property.StringProperty;
+import javafx.collections.FXCollections;
+import javafx.css.PseudoClass;
+import javafx.scene.Node;
+import javafx.scene.Parent;
+import javafx.scene.control.Labeled;
+
+import com.google.common.collect.ArrayListMultimap;
+
+class PreferencesSearchHandler {
+
+    private static PseudoClass labelHighlight = PseudoClass.getPseudoClass("search-highlight");
+    private final List<PrefsTab> preferenceTabs;
+    private final StringProperty searchText;
+    private final ListProperty<PrefsTab> filteredPreferenceTabs;
+    private final ArrayListMultimap<PrefsTab, LabeledWrapper> preferenceTabsLabelNames;
+    private final ArrayList<LabeledWrapper> highlightedLabels = new ArrayList<>();
+
+    private class LabeledWrapper {
+
+        private final Labeled labeled;
+
+        LabeledWrapper(Labeled _labeled) {
+            labeled = _labeled;
+        }
+
+        public boolean contains(String query) {
+            return labeled.getText().toLowerCase(Locale.ROOT).contains(query.toLowerCase(Locale.ROOT));
+        }
+
+        public String getText() {
+            return labeled.getText();
+        }
+
+        public void setHighlighted(boolean highlight) {
+            labeled.pseudoClassStateChanged(labelHighlight, highlight);
+        }
+    }
+
+    PreferencesSearchHandler(List<PrefsTab> preferenceTabs, StringProperty searchText) {
```

Comment:
Please remove the StringProperty as constructor argument, instead expose the property through a method:
And initialize the variable with `new SimpleStringProperty("")`
```
public StringProperty searchTextProperty()
{
return this.searchText;
 }
```

Reply:
As I understand, this should then be used bind to `searchBox.textProperty()`  in `PerferencesDialog`,
`searchHandler.searchTextProperty().bind(searchBox.textProperty());` ?

Also, shouldn't `       tabsList.itemsProperty().bindBidirectional(searchHandler.filteredPreferenceTabsProperty());`
have a unidirectional binding instead?

Reply:
@Siedlerchr I am facing some issues on implementing this. 
I have removed the `StringProperty` from the constructor argument as per your suggestion. 
In `PreferencesDialog`, the `searchHandler.searchStringProperty()` is bound to `searchBox.textProperty()`. However, when the `searchBox` text changes, it doesn't trigger the ChangeListener for `searchText`. Unless, I add a statement to listen for changes on `searchBox` as well.
```
 searchBox.textProperty().addListener((o, p, n) -> {
        // Do nothing.
});
```

I am not able to understand the reason for this, is it necessary to have listeners on both the properties which are bound?